packages/SwingSet/src/liveslots/virtualObjectManager.js

```diff
@@ -352,6 +357,53 @@ export function makeVirtualObjectManager(
     configurable: true,
   });
 
+  /**
+   * Assess the facetiousness of a value.  If the value is an object containing
+   * only named properties and each such property's value is a function, `obj`
+   * represents a single facet and 'one' is returned.  If each property's value
+   * is instead an object of facetiousness 'one', `obj` represents multiple
+   * facets and 'many' is returned.  In all other cases `obj` does not represent
+   * any kind of facet abstraction and 'not' is returned.
+   *
+   * @typedef {'one'|'many'|'not'} Facetiousness
+   *
+   * @param {*} obj  The (alleged) object to be assessed
+   *
+   * @returns {Facetiousness} an assessment of the facetiousness of `obj`
+   */
+  function assessFacetiousness(obj) {
```

Comment:
Protecting against `obj` being a cyclic structure is probably warranted paranoia.  That will be quite simple because the recursion is not allowed to go more than 1 layer deep.

`obj` can't change its shape while we're looking at it because we have control.

`defineKind` calls `Far` on the facets so the rest of the stuff you're concerned about here is not an issue.

Reply:
Unit testing here seems sensible but raises an issue that has always bugged me a bit about unit tests, namely that implementing them requires exposing APIs that are not supposed to be exposed.

Reply:
See my reasoning below for why it is fine for `assertFacetiousness` does not need to be defensive or validating. It operates on an unfrozen object. Trying to validate that there's anything stable about an unfrozen object is close enough to impossible. Instead, whatever decision it makes then needs to be enforced *after* the relevant objects are hardened, so that they'll hold still while we're examining them. The worst a TOCTTOU can do is cause the making of the representative to fail.

@FUDCo @warner please check my reasoning here. If you buy it, it is certainly worth a significant inline comment so they next reader doesn't have to recapitulate this reasoning.